### ISSUE TITLE
[bug] Videos API does not respect `local` parameter

### ISSUE THREAD

**Issue by Avoxel284** [iv-org/invidious#3221]
**Is your feature request related to a problem? Please describe.**
For _some videos_, any stream URL in `adaptiveFormats` (when using the API) returns with a HTTP 403 error, most likely due to IP reasons as seen in a similar bug (#3079). However, in #3079, they are using a different endpoint and not the API, which only gives the stream URL, not additional metadata like bitrate, audio quality, etc.

**Describe the solution you'd like**
From what I've seen in #3079, Invidious provides a `local=true` query parameter as a fix for this issue. If some kind of similar (or same) parameter is added to the API so all the stream URLs are localised, that'd be appreciated.

**Describe alternatives you've considered**
As I mentioned, I could use the endpoint used in #3079, but I need the metadata the API provides.

**Comment by SamantazFox** [comment]
You can pass `?local=true` to have the URLs transformed to use the instance as a proxy.

Note that trying to reach any `adaptiveFormats` URLs that points directly to google's servers will always result in a 403 error. Youtube is sending CORS headers that prevents your browser from acessing those URLs outside of `youtube.com`. By using the instance as a proxy, the headers can be stripped and the video played.

**Comment by Avoxel284** [comment]
Do you mean `?local=true` as in `https://vid.puffyan.us/api/v1/videos/videoid?local=true`?

**Comment by SamantazFox** [comment]
> Do you mean `?local=true` as in `https://vid.puffyan.us/api/v1/videos/videoid?local=true`?

yes

> Also, for most videos, the `adaptiveFormats` URLs that point to google servers do work

Oh, that's weird... @unixfox any clues why?

**Comment by unixfox** [comment]
> > Also, for most videos, the `adaptiveFormats` URLs that point to google servers do work
> 
> Oh, that's weird... @unixfox any clues why?

Obviously it will load, but in the list there are videos without audio. You can load them without any CORS issue but you won't have any audio.

That's why DASH exist, it combines the video URL and the audio URL for a specific youtube video.

**Comment by Avoxel284** [comment]
When I make a get request to https://vid.puffyan.us/api/v1/videos/2cWu_ow99Wk, **all** the `adaptiveFormats` URLs return normal audio and video.
However, when I make a get request to https://vid.puffyan.us/api/v1/videos/EUcvdGfGEj0, **all** the `adaptiveFormats` URLs return 403...
And ig ofc it also affects when not using the API, https://vid.puffyan/watch?v=EUcvdGfGEj0.

> > Do you mean `?local=true` as in `https://vid.puffyan.us/api/v1/videos/videoid?local=true`?
>
> yes

I tried adding the `?local=true` param on the API req and it didn't make a difference 🤔

**Comment by SamantazFox** [comment]
both videos you mentionned play well in both DASH and HD720 (proxied).

> I tried adding the `?local=true` param on the API req and it didn't make a difference 🤔

That shouldn't happen, yeah...

**Comment by Avoxel284** [comment]
If it `?local=true` is supposed to fix this, I'd also recommend including it in the API documentation 👍

**Comment by SamantazFox** [comment]
Yeah, I need to rewrite the whole API documentation at some point :c

**Comment by Avoxel284** [comment]
Just pointing out, I looked at the API code, and I could only find that the `id` and `region` params are taken and used from the `/api/v1/videos` route, not `local`...

Anyway, I'd like to reiterate the question; How would I be able to stream any `adaptiveFormats` URL without running into a 403 Forbidden error from Google, since I've been able to do so in the past for videos that are coincidentally less popular (meaning maybe no one has loaded the video). From code-digging, I think this may actually be an issue related to how videos are cached in the database (if this is right, then ig this should be considered a bug).

I also noticed a force refresh arg in `get_video` that bypasses getting cached videos from the db, so maybe add the feature to set this from the query params (tbh could contradict the purpose of caching in the first place, but would be useful for testing though)?
